File: packages/core/src/core/frontmatter.ts
import { isTag } from 'domhandler';
import type { Document, Element } from 'domhandler';
import { getTextContent } from '../rules/util.js';

/**
 * Patterns that require YAML quoting when present in a value.
 * Uses `: ` (colon-space) rather than bare `:` to avoid quoting URLs.
 */
const NEEDS_QUOTING = /: |[#"'{}[\]|>\n\\]|^\s|\s$/;

/**
 * Extracts page metadata from the `<head>` element of an HTML document.
 *
 * Extracts:
 * - `title` from `<title>` text content
 * - `description` from `<meta name="description">`
 * - `image` from `<meta property="og:image">`
 *
 * @param document - The parsed DOM document.
 * @returns A record containing only fields with non-empty values.
 */
export function extractMetadata(document: Document): Record<string, string> {
    const meta: Record<string, string> = {};

    const head = findHead(document);
    if (!head) return meta;

    for (const child of head.children) {
        if (!isTag(child)) continue;

        if (child.name === 'title') {
            const text = getTextContent(child).trim();
            if (text) meta.title = text;
        }

        if (child.name === 'meta') {
            const entry = extractFromMeta(child.attribs);
            if (entry) meta[entry[0]] = entry[1];
        }
    }

    return meta;
}

/**
 * Extracts a key-value pair from a `<meta>` element's attributes.
 *
 * Recognizes `<meta name="description">` and `<meta property="og:image">`.
 *
 * @param attrs - The element's attribute map.
 * @returns A `[key, value]` tuple, or `undefined` for unrecognized or empty meta elements.
 */
function extractFromMeta(attrs: Record<string, string>): [string, string] | undefined {
    if (!('content' in attrs)) return undefined;
    const content = attrs.content.trim();
    if (!content) return undefined;

    if ('name' in attrs && attrs.name.toLowerCase() === 'description') return ['description', content];
    if ('property' in attrs && attrs.property.toLowerCase() === 'og:image') return ['image', content];

    return undefined;
}

/**
 * Serializes a metadata record into a YAML frontmatter block.
 *
 * Output order: `title`, `description`, `image`, then remaining keys alphabetically.
 *
 * @param meta - The key-value metadata record to serialize.
 * @returns A `---`-delimited YAML frontmatter string, or an empty string if `meta` is empty.
 */
export function serializeFrontmatter(meta: Record<string, string>): string {
    const keys = Object.keys(meta);
    if (keys.length === 0) return '';

    const priority = ['title', 'description', 'image'];
    const rest = keys.filter(k => !priority.includes(k));
    rest.sort((a, b) => a.localeCompare(b));
    const ordered = [...priority.filter(k => k in meta), ...rest];

    const lines = ordered.map(key => `${key}: ${yamlQuote(meta[key])}`);
    return `---\n${lines.join('\n')}\n---\n`;
}

/**
 * Wraps a YAML value in double quotes if it contains characters that
 * require quoting (colons, hashes, brackets, leading/trailing whitespace, etc.).
 *
 * @param value - The raw YAML value string.
 * @returns The value as-is or wrapped in escaped double quotes.
 */
function yamlQuote(value: string): string {
    if (NEEDS_QUOTING.test(value)) {
        const escaped = value.replaceAll('\\', String.raw`\\`).replaceAll('"', String.raw`\"`);
        return `"${escaped}"`;
    }
    return value;
}

/**
 * Walks the document to find a `<head>` element.
 *
 * Handles both `<html><head>...` and bare `<head>` as a direct document child.
 *
 * @param document - The parsed DOM document.
 * @returns The `<head>` element, or `undefined` if not found.
 */
function findHead(document: Document): Element | undefined {
    for (const child of document.children) {
        if (!isTag(child)) continue;

        if (child.name === 'head') return child;

        if (child.name === 'html') {
            for (const grandchild of child.children) {
                if (isTag(grandchild) && grandchild.name === 'head') {
                    return grandchild;
                }
            }
        }
    }
    return undefined;
}

File: python/src/markdown_for_agents/core/frontmatter.py
from __future__ import annotations

import re

from ..dom.nodes import Document, Element, is_tag
from ..rules.util import get_text_content

_NEEDS_QUOTING = re.compile(r": |[#\"'{}[\]|>\n\\]|^\s|\s$")


def extract_metadata(document: Document) -> dict[str, str]:
    """Extract page metadata from the <head> element of an HTML document."""
    meta: dict[str, str] = {}

    head = _find_head(document)
    if head is None:
        return meta

    for child in head.children:
        if not is_tag(child):
            continue
        assert isinstance(child, Element)

        if child.name == "title":
            text = get_text_content(child).strip()
            if text:
                meta["title"] = text

        if child.name == "meta":
            entry = _extract_from_meta(child.attribs)
            if entry is not None:
                meta[entry[0]] = entry[1]

    return meta


def _extract_from_meta(attrs: dict[str, str]) -> tuple[str, str] | None:
    """Extract a key-value pair from a <meta> element's attributes."""
    if "content" not in attrs:
        return None
    content = attrs["content"].strip()
    if not content:
        return None

    if "name" in attrs and attrs["name"].lower() == "description":
        return ("description", content)
    if "property" in attrs and attrs["property"].lower() == "og:image":
        return ("image", content)

    return None


def serialize_frontmatter(meta: dict[str, str]) -> str:
    """Serialize a metadata record into a YAML frontmatter block."""
    keys = list(meta.keys())
    if not keys:
        return ""

    priority = ["title", "description", "image"]
    rest = sorted([k for k in keys if k not in priority])
    ordered = [k for k in priority if k in meta] + rest

    lines = [f"{key}: {_yaml_quote(meta[key])}" for key in ordered]
    return f"---\n{chr(10).join(lines)}\n---\n"


def _yaml_quote(value: str) -> str:
    """Wrap a YAML value in double quotes if it contains characters that require quoting."""
    if _NEEDS_QUOTING.search(value):
        escaped = value.replace("\\", "\\\\").replace('"', '\\"')
        return f'"{escaped}"'
    return value


def _find_head(document: Document) -> Element | None:
    """Walk the document to find a <head> element."""
    for child in document.children:
        if not is_tag(child):
            continue
        assert isinstance(child, Element)

        if child.name == "head":
            return child

        if child.name == "html":
            for grandchild in child.children:
                if is_tag(grandchild):
                    assert isinstance(grandchild, Element)
                    if grandchild.name == "head":
                        return grandchild

    return None
